File: Top 100 Codes/01-Getting Started/21-Finding Prime Factors of a number/FindingPrimeFactorsofanumber.py
# Function to find the prime factors of a number using recursion
def Prime_Factorial(n, arr):
    # If the number is less than 4, it's either prime or 1
    if n < 4:
        arr.append(n)
        return arr

    # Iterate from 2 to half of the number
    for i in range(2, 1 + n // 2):
        # If we reach the middle of the range without finding factors, 'n' is prime
        if i == (1 + n // 2):
            arr.append(n)
            return arr

        # If 'i' divides 'n', it's a factor
        if n % i == 0:
            arr.append(i)  # Append the factor to the list
            n = n // i    # Divide 'n' by the factor
            return Prime_Factorial(n, arr)  # Recursively find the factors

    return arr

# Main program to take user input
n = int(input("Enter a number to find its prime factors: "))
arr = []  # List to store the prime factors

# Validate input to ensure it's positive
if n <= 0:
    print("Please enter a positive integer.")
else:
    # Find and display the prime factors
    factors = Prime_Factorial(n, arr)
    print(f"Prime factors of {n} are: {factors}")

File: Top 100 Codes/01-Getting Started/21-Finding Prime Factors of a number/FindingPrimeFactorsofanumber.cpp
#include <iostream>
#include <cmath>
using namespace std;

// Function to print all prime factors of a given number
void primeFactors(int n) {
    // Handle the number of 2s that divide n
    while (n % 2 == 0) {
        cout << 2 << " ";
        n /= 2;
    }

    // Handle odd factors from 3 to sqrt(n)
    for (int i = 3; i <= sqrt(n); i += 2) {
        while (n % i == 0) {
            cout << i << " ";
            n /= i;
        }
    }

    // If n is a prime number greater than 2
    if (n > 2)
        cout << n << " ";
}

int main() {
    int n;

    // Prompt user for input
    cout << "Enter a number to find its prime factors: ";
    cin >> n;

    // Validate input to ensure it's positive
    if (n <= 0) {
        cout << "Please enter a positive integer." << endl;
        return 1; // Exit with an error code
    }

    // Print the prime factors of the number
    cout << "Prime factors of " << n << " are: ";
    primeFactors(n);
    cout << endl;

    return 0;
}

// Time Complexity: O(sqrt(N)) - Iterates up to sqrt(n) for finding factors
// Space Complexity: O(1) - Constant space is used
